aim/agent/aid/universes/aci/aci_universe.py

```diff
@@ -113,37 +124,152 @@ def _reload_websocket_config(self):
         if set(ws_urls) != set(self.ws_urls):
             self.ws_urls = ws_urls
 
+    def _update_apic_assign_db(self, aim_context, apic_assign,
+                               apic_assign_obj):
+        try:
+            if apic_assign_obj is None:
+                apic_assign.aim_aid_id = self.agent_id
+                return self.manager.create(aim_context, apic_assign)
+
+            if apic_assign_obj.aim_aid_id != self.agent_id:
+                obj = self.manager.update(aim_context, apic_assign_obj,
+                                          aim_aid_id=self.agent_id)
+            else:
+                # This will update the last_update_timestamp
+                # automatically
+                obj = self.manager.update(aim_context, apic_assign_obj)
+            return obj
+        # This means another controller is also adding/updating
+        # the same entry at the same time and he has beat us.
+        except Exception as e:
+            LOG.info(e)
+            return None
+
+    def _ws_session_login(self, url, url_max_retries, purpose):
+        retries = 0
+        LOG.info('Establishing %s WS connection with url: %s',
+                 purpose, url)
+        while retries < url_max_retries:
+            session = acitoolkit.Session(
+                url, self.apic_username,
+                self.apic_password,
+                verify_ssl=self.verify_ssl_certificate,
+                cert_name=self.cert_name,
+                key=self.private_key_file)
+            resp = session.login()
+            if not resp.ok:
+                LOG.warn('%s Websocket connection failed: %s',
+                         purpose, resp.text)
+                retries += 1
+                if session.session:
+                    session.close()
+                continue
+            LOG.info('%s Websocket connection succeeded with url: %s',
+                     purpose, url)
+            return session
+
+        return None
+
+    def establish_ws_session_for_recovery(self, max_retries=None):
```

Comment:
This function looks a lot like establish_ws_session() below. Could they be folded into a single function that handles both "normal" and "recovery" situations? Also, I'm still not really clear on why recovery needs to be handled differently.

Reply:
The design in this patch is that we use the new table to indicate who has claimed the ownership of an APIC already. If an APIC is already claimed, aim-aid will try to look for the next APIC on which nobody has laid the claim yet. If it couldn't find such an APIC, it will then go ahead to use 1 APIC which already has an owner then enter the recovery mode. When in the recovery mode, this aim-aid will try to find the next free APIC periodically with a back-off timer. Recovery needs to be handled slightly different because if this aim-aid still couldn't locate a free APIC, its totally fine (because currently its already sharing an APIC with some other owner) so no need to re-connect if that happens.

Reply:
I was actually thinking about using one single function to do it but didn't proceed further as I think code will get a bit messy with quite a few if-else statement here & there. In general I like the idea of avoiding code duplication so I could take another look then.

Reply:
Assuming that the "full-mesh" approach is not workable or would require too much effort to implement, I guess the fundamental concern I have is whether a single AID having "ownership" of an APIC is the right concept to implement. This makes sense only if we assume there are always at least as many APICs as there are AIDs in a healthy deployment. Then, as currently implemented in this patch, the normal situation is for at most one AID to "own" each APIC, and temporary loss of one or more APICs can be thought of as "recovery".

I understand ACI requires at least 3 APICs, and that 3 APICs is a typical deployment configuration. So, is deploying more than 3 AIDs useful? If deploying multiple AIDs is intended only for availability, then 3 AIDs is probably sufficient. If 3 AIDs isn't considered sufficient redundancy, then 3 APICs probably isn't sufficient either, so requiring at least as many APICs as AIDs may be reasonable. But if additional AIDs are needed for scalability, and not just for availability, then more AIDs than APICs becomes a normal situation we should support. This is where the concept of at most one AID owning an APIC becomes problematic. 

If we do need to optimally support configurations with more AIDs than APICs, we could do this by replacing the concept of "ownership" here with one of "association". The DB table would use the AID identity instead of the APIC identity as the primary key, and a row would record which APIC each AID is currently using. Each AID would need to periodically update its row in the table so timestamps would allow rows that have not recently been updated to be ignored. When connecting, an AID would query the whole table, and then randomly pick from the APICs used by the least number of AIDs. I don't think this "association" approach would require significantly more effort to implement than the "ownership" approach used in this patch, but I could be missing something.

If we are convinced that we definitely do not need to (optimally) support configurations with more AIDs than APICs, then my concern with this patch was the code duplication between "normal" and "recovery" cases. The normal case code, _establish_ws_connection(), would be executed all the time, and any bugs in it would be apparent and get fixed. But  _establish_ws_connection_for_recovery() would rarely be executed in production. Any future fixes or enhancements to one would need to be manually applied to the other, or they'd diverge over time. I'll take closer look to see if this has been adequately addressed, or if additional refactoring might be helpful.

Reply:
My understanding is that most production deployments would have 3 openStack controllers and 3 APICs. In some productions we could even have more APICs up to 7. So its actually rare to have more openStack controllers than APICs. But even if that happens, this patch can handle that well. The extra AIDs will go into recovery mode with back-off timer which eventually will only try to look for free APIC in every 10 mins so it won't cause a huge burden on the system.

The problem of the 'association' approach is that then every AID will have to periodically scan the whole table to re-balance the load for a use case like an APIC could go down then come back up again. While in the 'owndership" solution, only the extra AID will need to be in the recovery mode.

For the duplication code concern, I've consolidated those so there is no more duplication code now. Thx.